mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowHttpCaller.java

```diff
@@ -30,20 +30,65 @@
 class MlflowHttpCaller {
   private static final Logger logger = LoggerFactory.getLogger(MlflowHttpCaller.class);
   private static final String BASE_API_PATH = "api/2.0/preview/mlflow";
-  private HttpClient httpClient;
+  protected HttpClient httpClient;
   private final MlflowHostCredsProvider hostCredsProvider;
+  private final int maxRateLimitIntervalMillis;
+  private final int rateLimitRetrySleepInitMillis;
 
   MlflowHttpCaller(MlflowHostCredsProvider hostCredsProvider) {
+    this(hostCredsProvider, 60000);
+  }
+
+  MlflowHttpCaller(MlflowHostCredsProvider hostCredsProvider, int maxRateLimitIntervalSeconds) {
+    this(hostCredsProvider, maxRateLimitIntervalSeconds, 1000);
+  }
+
+  MlflowHttpCaller(MlflowHostCredsProvider hostCredsProvider,
+                   int maxRateLimitIntervalSeconds,
+                   int rateLimitRetrySleepInitMs) {
     this.hostCredsProvider = hostCredsProvider;
+    this.maxRateLimitIntervalMillis = maxRateLimitIntervalSeconds;
+    this.rateLimitRetrySleepInitMillis = rateLimitRetrySleepInitMs;
+  }
+
+  MlflowHttpCaller(MlflowHostCredsProvider hostCredsProvider,
+                   int maxRateLimitIntervalSeconds,
+                   int rateLimitRetrySleepInitMs,
+                   HttpClient client) {
+    this(hostCredsProvider, maxRateLimitIntervalSeconds, rateLimitRetrySleepInitMs);
+    this.httpClient = client;
+  }
+
+
+  HttpResponse executeRequest(HttpRequestBase request) throws IOException {
+    HttpResponse response = httpClient.execute(request);
+    int timeLeft = maxRateLimitIntervalMillis;
+    int sleepFor = rateLimitRetrySleepInitMillis;
+    while (response.getStatusLine().getStatusCode() == 429 && timeLeft > 0) {
+      timeLeft -= sleepFor;
+      logger.warn("Request returned with status code 429 (Rate limit exceeded). Retrying after "
+                  + sleepFor
+                  + " milliseconds. Will continue to retry 429s for up to "
+                  + timeLeft
```

Comment:
this total wait time (`timeLeft`) includes sleep time but message does not say so.

Reply:
What message would you suggest? 

The time_left only includes sleep time. I did not want to complicate the message since I don't think it needs to be super precise and the assumption was that 429s are fast.

Reply:
After our discussion offline, I have moved the timeLeft update past the warning printout.